File: array and string/anagram.js
/*
Write a function, anagrams, that takes in two strings as arguments. The function should return a boolean indicating whether or not the strings are anagrams. Anagrams are strings that contain the same characters, but in any order.

test_00:

anagrams('restful', 'fluster'); // -> true
test_01:

anagrams('cats', 'tocs'); // -> false
test_02:

anagrams('monkeyswrite', 'newyorktimes'); // -> true
test_03:

anagrams('paper', 'reapa'); // -> false
test_04:

anagrams('elbow', 'below'); // -> true
test_05:

anagrams('tax', 'taxi'); // -> false
test_06:

anagrams('taxi', 'tax'); // -> false
test_07:

anagrams('night', 'thing'); // -> true
test_08:

anagrams('abbc', 'aabc'); // -> false
test_09:

anagrams('po', 'popp'); // -> false
test_10:

anagrams('pp', 'oo') // -> false
*/

const anagrams = (s1, s2) => {
    // todo
    if (s1.length !== s2.length) return false
    
    let obj1 = getObject(s1)
    let obj2 = getObject(s2)
   
    for(let key in obj1){
        if(obj1[key] !== obj2[key] || obj2[key] === undefined) return false
    }
      return true
  };
  
  function getObject(str){
      let obj = {}
       for(let i = 0 ; i < str.length; i++){
             if(obj[str[i]]){
                  obj[str[i]] += 1
             }else{
                  obj[str[i]] = 1
             }
       }
      return obj
  }

File: array and string/anagram.py
# anagrams
# Write a function, anagrams, that takes in two strings as arguments. The function should return a boolean
#  indicating whether or not the strings are anagrams. Anagrams are strings that contain the same characters, but in any order.

# anagrams('restful', 'fluster') # -> True
# anagrams('cats', 'tocs') # -> False
# anagrams('monkeyswrite', 'newyorktimes') # -> True
# anagrams('paper', 'reapa') # -> False
# anagrams('elbow', 'below') # -> True
# anagrams('tax', 'taxi') # -> False
# anagrams('taxi', 'tax') # -> False
# anagrams('night', 'thing') # -> True
# anagrams('abbc', 'aabc') # -> False
# anagrams('po', 'popp') # -> false
# anagrams('pp', 'oo') # -> false

def anagrams(s1, s2):
    hash = makeHash(s1)
    for el in s2:
        if hash.get(el, 0) == 0:
            return False
        hash[el] -= 1
    all_zero = all(value == 0 for value in hash.values())
    return all_zero

def makeHash(s1):
    hash = {}
    for el in s1:
        hash[el] = hash.get(el, 0) + 1
    return hash